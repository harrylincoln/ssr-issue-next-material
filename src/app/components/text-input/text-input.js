import ErrorIcon from '../error-icon/error-icon';
import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';
import styles from './text-input.styles';
import { withStyles } from '@material-ui/core/styles';

export const SimpleTextInput = ({
    classes,
    input,
    label,
    placeholder,
    autocomplete,
    disabled,
    dataSelector,
    readOnly, // TODO: implement readonly styles
    type,
    name,
    meta: { touched, error, warning },
}) => {
    const classMap = cx({
        [classes.formField]: true,
        [classes.error]: touched && (error || warning),
    });

    const inputProps = {
        disabled,
        readOnly,
        placeholder,
        type,
        ...input,
    };

    return (
        <div className={classMap}>
            {type === 'checkbox' ? (
                <label htmlFor={name}>
                    <input
                        data-selector={dataSelector}
                        className={`${classes.formElement} ${classes.input}`}
                        {...inputProps}
                    />
                    <span className={classes.formLabel}>{label}</span>
                </label>
            ) : (
                <label htmlFor={name}>
                    <span className={classes.formLabel}>{label}</span>
                    <input
                        autoComplete={autocomplete}
                        data-selector={dataSelector}
                        className={classes.formElement}
                        {...inputProps}
                    />
                </label>
            )}
            {touched &&
                ((error && (
                    <span className={classes.formWarning}>
                        <ErrorIcon />
                        {error}
                    </span>
                )) ||
                    (warning && <span className="warning">{warning}</span>))}
        </div>
    );
};

SimpleTextInput.propTypes = {
    classes: PropTypes.shape({}),
    disabled: PropTypes.bool,
    input: PropTypes.shape(),
    label: PropTypes.string.isRequired,
    type: PropTypes.oneOf(['email', 'text', 'tel', 'checkbox', 'password']),
    placeholder: PropTypes.string,
    dataSelector: PropTypes.string,
    name: PropTypes.string,
    readOnly: PropTypes.bool,
    meta: PropTypes.shape({
        touched: PropTypes.bool,
        warning: PropTypes.string,
        error: PropTypes.string,
    }),
    autocomplete: PropTypes.string,
};

SimpleTextInput.defaultProps = {
    classes: {},
    input: {},
    autocomplete: null,
    disabled: false,
    name: null,
    readOnly: false,
    placeholder: undefined,
    type: 'text',
    dataSelector: '',
    meta: {},
};

export default withStyles(styles)(SimpleTextInput);
