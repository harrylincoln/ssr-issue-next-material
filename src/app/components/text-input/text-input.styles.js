const textInputStyles = (theme) => ({
    ...theme.pallete.primary.atomics,
    ...theme.pallete.primary.forms,
    ...theme.pallete.primary.generalSelectorClasses,

    error: {
        '& $formElement': {
            ...theme.pallete.primary.forms.formElementError,
        },
    },
});

export default textInputStyles;
