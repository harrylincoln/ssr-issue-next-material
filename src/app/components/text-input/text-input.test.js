import ErrorIcon from '../error-icon/error-icon';
import React from 'react';
import { SimpleTextInput as TextInput } from './text-input';
import { shallow } from 'enzyme';

describe('TextInput', () => {
    const error = 'error';
    const warning = 'warning';
    const touched = true;
    const spyScrollTo = jest.fn();

    beforeEach(() => {
        Object.defineProperty(global.window, 'scrollTo', {
            value: spyScrollTo,
        });
        Object.defineProperty(global.window, 'scrollY', { value: 1 });
        spyScrollTo.mockClear();
    });

    it('should render correctly', () => {
        const component = shallow(<TextInput name="name" label="label" />);
        expect(component).toMatchSnapshot();
    });

    it('should render error message', () => {
        const component = shallow(
            <TextInput
                touched
                meta={{ error, touched }}
                name="name"
                label="label"
                classes={{ formWarning: 'formWarning' }}
            />
        );
        expect(
            component.contains(
                <span className="formWarning">
                    <ErrorIcon />
                    {error}
                </span>
            )
        ).toEqual(true);
    });

    it('should render warning', () => {
        const component = shallow(
            <TextInput
                touched
                meta={{ warning, touched }}
                name="name"
                label="label"
            />
        );
        expect(
            component.contains(<span className="warning">{warning}</span>)
        ).toEqual(true);
    });

    describe('input callbacks', () => {
        const input = {
            onChange: jest.fn(),
            onBlur: jest.fn(),
            onFocus: jest.fn(),
        };

        afterEach(() => {
            input.onChange.mockClear();
            input.onBlur.mockClear();
            input.onFocus.mockClear();
        });
        it('should call onFocus for focus event', () => {
            const component = shallow(
                <TextInput name="name" label="label" input={input} />
            );
            const textInput = component.find('input');

            textInput.simulate('focus');
            expect(input.onFocus).toHaveBeenCalledTimes(1);
        });
        it('should call onChange for change event', () => {
            const component = shallow(
                <TextInput name="name" label="label" input={input} />
            );
            const textInput = component.find('input');

            textInput.simulate('change', { a: 'b' });
            expect(input.onChange).toHaveBeenCalledTimes(1);
            expect(input.onChange).toHaveBeenCalledWith({ a: 'b' });
        });

        it('should call onBlur for blur event', () => {
            const component = shallow(
                <TextInput name="name" label="label" input={input} />
            );
            const textInput = component.find('input');

            textInput.simulate('blur');
            expect(input.onBlur).toHaveBeenCalledTimes(1);
        });
    });
});
