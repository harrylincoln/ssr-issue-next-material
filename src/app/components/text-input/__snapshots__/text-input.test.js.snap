// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput should render correctly 1`] = `
<div
  className=""
>
  <label
    htmlFor="name"
  >
    <span>
      label
    </span>
    <input
      autoComplete={null}
      data-selector=""
      disabled={false}
      readOnly={false}
      type="text"
    />
  </label>
</div>
`;
